### PR TITLE
xy <-> yx for bboxes 

### PR DESCRIPTION
**Pull request by yuyu2172**
A code I used to convert from the data for voc_detection_dataset on the web to the new one.

```python
from six.moves.urllib import request
import numpy as np
import os


fns = ['voc_detection_result_2007_test_truncated_2017_06_02.npz',
       'voc_detection_dataset_2007_test_truncated_2017_06_02.npz']
out_fns = ['voc_detection_result_2007_test_truncated_2017_06_06.npz',
           'voc_detection_dataset_2007_test_truncated_2017_06_06.npz']

flags = [True, False]

for fn, out_fn, flag in zip(fns, out_fns, flags):
    base_url = 'https://github.com/yuyu2172/' \
        'share-weights/releases/download/0.0.2'
    result = np.load(request.urlretrieve(os.path.join(
        base_url,
        fn))[0])

    new_bboxes = []
    for bbox in result['bboxes']:
        old_x_min = bbox[:, 0]
        old_y_min = bbox[:, 1]
        old_x_max = bbox[:, 2]
        old_y_max = bbox[:, 3]
        new_bboxes.append(
            np.stack((old_y_min, old_x_min, old_y_max, old_x_max), axis=1))
    if flag:
        np.savez(
            out_fn,
            bboxes=result['bboxes'], labels=result['labels'], scores=result['scores'])
    else:
        np.savez(
            out_fn,
            bboxes=result['bboxes'], labels=result['labels'],
            difficults=result['difficults'])

```

EDIT:

```python
from six.moves.urllib import request
import numpy as np
import os


fns = ['voc_detection_result_2007_test_truncated_2017_06_02.npz',
       'voc_detection_dataset_2007_test_truncated_2017_06_02.npz']
out_fns = ['voc_detection_result_2007_test_truncated_2017_06_06.npz',
           'voc_detection_dataset_2007_test_truncated_2017_06_06.npz']

flags = [True, False]

for fn, out_fn, flag in zip(fns, out_fns, flags):
    base_url = 'https://github.com/yuyu2172/' \
        'share-weights/releases/download/0.0.2'
    result = np.load(request.urlretrieve(os.path.join(
        base_url,
        fn))[0])

    new_bboxes = []
    for bbox in result['bboxes']:
        old_x_min = bbox[:, 0]
        old_y_min = bbox[:, 1]
        old_x_max = bbox[:, 2]
        old_y_max = bbox[:, 3]
        new_bboxes.append(
            np.stack((old_y_min, old_x_min, old_y_max, old_x_max), axis=1))
    if flag:
        np.savez(
            out_fn,
            bboxes=new_bboxes, labels=result['labels'], scores=result['scores'])
    else:
        np.savez(
            out_fn,
            bboxes=new_bboxes, labels=result['labels'],
            difficults=result['difficults'])

```

### CODE REVIEW

**Comment by Hakuyume** [comment]
```
    if flag:
        np.savez(
            out_fn,
            bboxes=result['bboxes'], labels=result['labels'], scores=result['scores'])
    else:
        np.savez(
            out_fn,
            bboxes=result['bboxes'], labels=result['labels'],
            difficults=result['difficults'])
```
Use `new_bboxes`.

File: chainercv/utils/bbox/non_maximum_suppression.py [hunk 1]
```diff
@@ -127,10 +127,10 @@ def _non_maximum_suppression_gpu(bbox, thresh, score=None, limit=None):
...
-  float left = max(bbox_a[0], bbox_b[0]);
-  float right = min(bbox_a[2], bbox_b[2]);
-  float top = max(bbox_a[1], bbox_b[1]);
-  float bottom = min(bbox_a[3], bbox_b[3]);
+  float top = max(bbox_a[0], bbox_b[0]);
+  float bottom = min(bbox_a[2], bbox_b[2]);
+  float left = max(bbox_a[1], bbox_b[1]);
+  float right = min(bbox_a[3], bbox_b[3]);
   float width = max(right - left, 0.f);
   float height = max(bottom - top, 0.f);
```
Comment (Hakuyume) [inline chainercv/utils/bbox/non_maximum_suppression.py]: `height`, `width` (Sorry, I missed it in review of #234)

File: chainercv/utils/bbox/non_maximum_suppression.py [hunk 2]
```diff
@@ -127,10 +127,10 @@ def _non_maximum_suppression_gpu(bbox, thresh, score=None, limit=None):
...
   float area_i = width * height;
```
Comment (Hakuyume) [inline chainercv/utils/bbox/non_maximum_suppression.py]: `height * width`

**Comment by yuyu2172** [comment]
```python
import chainer
from chainercv.links import FasterRCNNVGG16
from chainercv.links import SSD300
from chainercv.links import SSD512
from six.moves.urllib import request
import os
import numpy as np


def convert_xy_linear(l):
    l = l.copy()
    b = l.b.data.reshape(-1, 4)
    b_old_x = b[:, [0, 2]]
    b_old_y = b[:, [1, 3]]
    b[:, [0, 2]] = b_old_y
    b[:, [1, 3]] = b_old_x

    out_C, in_C = l.W.shape
    W = l.W.data.reshape(-1, 4, in_C)
    W_old_x = W[:, [0, 2]]
    W_old_y = W[:, [1, 3]]
    W[:, [0, 2]] = W_old_y
    W[:, [1, 3]] = W_old_x

    l.b.data[:] = b.reshape(-1)
    l.W.data[:] = W.reshape(-1, in_C)
    return l


def convert_xy_conv(l):
    l = l.copy()
    b = l.b.data.reshape(-1, 4)
    b_old_x = b[:, [0, 2]]
    b_old_y = b[:, [1, 3]]
    b[:, [0, 2]] = b_old_y
    b[:, [1, 3]] = b_old_x

    out_C, in_C, kh, kw = l.W.shape
    W = l.W.data.reshape(-1, 4, in_C, kh, kw)
    W_old_x = W[:, [0, 2]]
    W_old_y = W[:, [1, 3]]
    W[:, [0, 2]] = W_old_y
    W[:, [1, 3]] = W_old_x

    l.b.data[:] = b.reshape(-1)
    l.W.data[:] = W.reshape(-1, in_C, kh, kw)
    return l


def convert_faster_rcnn():
    fn_path =\
        os.path.expanduser('~/.chainer/dataset/pfnet/chainercv/models/faster_rcnn_vgg16_voc07_2017_05_24.npz')

    model = FasterRCNNVGG16(n_fg_class=20)

    chainer.serializers.load_npz(fn_path, model)
    model.head.cls_loc = convert_xy_linear(model.head.cls_loc)
    model.rpn.loc = convert_xy_conv(model.rpn.loc)
    chainer.serializers.save_npz('faster_rcnn_vgg16_voc07_2017_06_06.npz', model)


def convert_ssd_300():
    fn_path =\
        os.path.expanduser('~/.chainer/dataset/pfnet/chainercv/models/ssd300_voc0712_2017_05_24.npz')
    model = SSD300(n_fg_class=20)

    chainer.serializers.load_npz(fn_path, model)

    loc = chainer.ChainList()
    for i in range(len(model.multibox.loc)):
        loc.add_link(convert_xy_conv(model.multibox.loc[i]))
    model.multibox.loc = loc

    chainer.serializers.save_npz('ssd300_voc0712_2017_06_06.npz', model)


def convert_ssd_512():
    fn_path =\
        os.path.expanduser('~/.chainer/dataset/pfnet/chainercv/models/ssd512_voc0712_2017_05_24.npz')
    model = SSD512(n_fg_class=20)

    chainer.serializers.load_npz(fn_path, model)

    loc = chainer.ChainList()
    for i in range(len(model.multibox.loc)):
        loc.add_link(convert_xy_conv(model.multibox.loc[i]))
    model.multibox.loc = loc

    chainer.serializers.save_npz('ssd512_voc0712_2017_06_06.npz', model)


if __name__ == '__main__':
    convert_ssd_512()
    convert_ssd_300()
    convert_faster_rcnn()
```

**Comment by yuyu2172** [comment]
I retrained Faster-RCNN and it scored 70.0mAP.

**Comment by Hakuyume** [comment]
`lt, rb` -> `tl, br`
https://github.com/yuyu2172/chainercv/blob/18fcf40889f72030d4a8b88f30d77c28da73ee01/chainercv/utils/bbox/bbox_iou.py#L35-L38
https://github.com/yuyu2172/chainercv/blob/18fcf40889f72030d4a8b88f30d77c28da73ee01/chainercv/utils/bbox/non_maximum_suppression.py#L83-L84

**Comment by Hakuyume** [comment]
Could you fix `examples/ssd/caffe2npz.py`?

File: examples/ssd/caffe2npz.py [hunk 3]
```diff
@@ -67,6 +67,24 @@ def _skip_layer(self, _):
         pass
 
 
+def convert_xy_conv(l):
+    b = l.b.data.reshape(-1, 4)
+    b_old_x = b[:, [0, 2]]
+    b_old_y = b[:, [1, 3]]
+    b[:, [0, 2]] = b_old_y
+    b[:, [1, 3]] = b_old_x
```
Comment (Hakuyume) [inline examples/ssd/caffe2npz.py]: How about `b = b[:, [1, 0, 3, 2]]`?

File: examples/ssd/caffe2npz.py [hunk 4]
```diff
@@ -100,8 +94,8 @@ def main():
     # The pretrained model outputs coordinates in xy convention.
     # This needs to be changed to yx convention, which is used
     # in ChainerCV.
-    for name in sorted([child.name for child in model.children()]):
-        if name[:12] == 'multibox/loc':
+    for name in [child.name for child in model.children()]:
+        if name.startswith('multibox/loc'):
```
Comment (Hakuyume) [inline examples/ssd/caffe2npz.py]:
```
for child in model.children():
    if child.name.startswith('multibox/loc'):
        convert_xy_conv(model[child.name])
```

**Pull request merged**